### DISCUSSION TITLE
Why doesn't pytest use decorators for tests discovery and/or hooks?

### DISCUSSION THREAD

**Discussion by alexisml** [pytest-dev/pytest#10783] [General] [answered]
In the early days of jUnit, the way the framework discovered and executed tests was by classes named `TestX` and methods called `testXYZ` (jUnit <= 3). That actually changed years ago (jUnit 4+), in favor of an explicit way of setting methods as tests: a decorator `@Test`.

Pytest does exactly what jUnit did way ago... why hasn't pytest adopted an explicit way of discovering tests (using a decorator)? 
Same comment applies for setup/teardown methods (which I have personally experiencied issues with importing some method from a class that was called `setup` and then was recognized as the actual setup of the test by pytest...), as well as hooks (which I personally hate because of being so obscure). 

Thanks

**Comment by nicoddemus** [comment, marked as answer]
> Pytest does exactly what jUnit did way ago... why hasn't pytest adopted an explicit way of discovering tests (using a decorator)?

Mostly backward compatibility reasons. Using decorators is supported by https://github.com/pytest-dev/pluggy, and `tox` for example started adopting hooks using decorators directly, but for pytest we never discussed adopting decorators as there has no user demand, and adopting an alternative way of doing things is always extra work.
